core-api/src/test/java/com/optimizely/ab/bucketing/BucketerTest.java

```diff
@@ -440,6 +447,153 @@ public void bucketUserNotInOverlappingGroupExperiment() throws Exception {
+
+        assertThat(algorithm.bucket(groupExperiment, "blah"),  is(variation));
+
+        logbackVerifier.expectMessage(Level.INFO,
+                "Returning previously activated variation \"e2_vtag1\" from user experiment registry.");
+
+        verify(userExperimentRecord).lookup("blah", groupExperiment.getKey());
+    }
+
+    @Test public void bucketUserRestoreActivationNullWithPersistentBucketer() throws Exception {
+        final AtomicInteger bucketValue = new AtomicInteger();
+        UserExperimentRecord userExperimentRecord = mock(UserExperimentRecord.class);
+        Bucketer algorithm = mockPersistentBucketAlgorithm(bucketValue, userExperimentRecord);
+        bucketValue.set(3000);
+
+        ProjectConfig projectConfig = validProjectConfig();
+        List<Experiment> groupExperiments = projectConfig.getGroups().get(0).getExperiments();
+        Experiment groupExperiment = groupExperiments.get(0);
+        final Variation variation = groupExperiment.getVariations().get(0);
+
+        when(userExperimentRecord.lookup("blah", groupExperiment.getKey())).thenReturn(null);
+
+        assertThat(algorithm.bucket(groupExperiment, "blah"),  is(variation));
+
+        logbackVerifier.expectMessage(Level.INFO,
+                "No previously activated variation returned from persistent bucketer.");
+
+        verify(userExperimentRecord).lookup("blah", groupExperiment.getKey());
+    }
+
+    @Test
+    public void nullUserExperimentRecordWhenCleaning() {
+        final AtomicInteger bucketValue = new AtomicInteger();
+        Bucketer algorithm = mockBucketAlgorithm(bucketValue);
+        bucketValue.set(3000);
+        try {
+            algorithm.cleanUserExperimentRecord();
+        } catch (NullPointerException e) {
+            fail();
+        }
+    }
+
+    @Test
+    public void nullUserExperimentRecords() {
+        final AtomicInteger bucketValue = new AtomicInteger();
+        UserExperimentRecord userExperimentRecord = mock(UserExperimentRecord.class);
+        Bucketer algorithm = mockPersistentBucketAlgorithm(bucketValue, userExperimentRecord);
+        bucketValue.set(3000);
+
+        when(userExperimentRecord.records()).thenReturn(null);
+        try {
+            algorithm.cleanUserExperimentRecord();
+        } catch (NullPointerException e) {
+            fail();
+        }
+    }
+
+    @Test
+    public void cleanRemovesRecordsOfExperimentsThatNoLongerExist() {
+        final AtomicInteger bucketValue = new AtomicInteger();
+        UserExperimentRecord userExperimentRecord = mock(UserExperimentRecord.class);
+        Bucketer algorithm = mockPersistentBucketAlgorithm(bucketValue, userExperimentRecord);
+        bucketValue.set(3000);
+
+        Map<String,Map<String,String>> records = new HashMap<String, Map<String, String>>();
+        Map<String,String> activation = new HashMap<String, String>();
+        activation.put("57", "57");
```

Comment:
this should be a mapping of experiment key to variation key, so to make it easier to read, use a more realistic mapping.